carnival/carnival.ts

```diff
@@ -0,0 +1,1017 @@
+                    scoreColor,
+                    image.font8
+                );
+
+                if (this.newBest == true) {
+                    this.image.printCenter(
+                        "New Best Score!",
+                        bestScoreHeight,
+                        scoreColor,
+                        image.font5
+                    );
+                } else {
+                    this.image.printCenter(
+                        "Best:" + this.best,
+                        bestScoreHeight,
+                        scoreColor,
+                        image.font8
+                    );
+                }
+            }
+        }
+    }
+
+    // End Info
+
+    // Formerly namespace Game
+
+
+    /**
+     * Adds additional end game styles
+     */
+    //% color="#b70082"
+    //% group="Game"
+    //% blockId=on_game_over_expanded
+    //% block="game over $winStyle || add effect $winEffect"
+    //% winType.defl=winTypes.Win
+    //% winEffect.defl=effects.confetti
+    //% inlineInputMode=inline
+    //% help=github:carnival/docs/on_game_over_expanded
+    export function onGameOverExpanded(winStyle: winTypes, winEffect?: effects.BackgroundEffect) {
+
+        if (winEffect == undefined) {
+            if (winStyle == winTypes.Lose) { winEffect = effects.melt; }
+            else { winEffect = effects.confetti; }
+        }
+
+        if (winStyle == winTypes.Lose) {
+            game.over(false, winEffect)
+        } else {
+            newGameOver(winStyle, winEffect);
+        }
+    }
+
+    /**
+     * Adds custom end game styles
+     */
+    //% color="#b70082"
+    //% group="Game"
+    //% blockId=on_game_over_custom_expanded
+    //% block="game over $message || with $winEffect and $gameSound send score $score judge $scoring"
+    //% message.defl="Great Job!"
+    //% scoring.defl=scoreTypes.None
+    //% winEffect.defl=effects.confetti
+    //% gameSound.defl=music.powerUp
+    //% inlineInputMode=inline
+    //% help=github:carnival/docs/on_game_over_custom_expanded
+    export function customGameOverExpanded(message: string, winEffect?: effects.BackgroundEffect, gameSound?: music.Melody, scoring?: scoreTypes, score?: number) {
+        if (!winEffect) { winEffect = effects.confetti; }
+        if (!scoring) { scoring = scoreTypes.HScore; }
+        if (score == undefined) { info.score(); }
+        if (!gameSound) { gameSound = music.powerUp; }
+        game.setGameOverSound(true, gameSound);
+        newGameOver(winTypes.Custom, winEffect, gameSound, scoring, message, score);
+    }
+
+
+    // Formerly namespace Ball
+
+
+
+    /**
+     * Creates a new throwable from an image and kind
+     * @param img the image for the sprite
+     * @param kind the kind to make the throwable
+     * @param x optional initial x position, eg: 10
+     * @param y optional initial y position, eg: 110
+     */
+    //% blockId=throw_create block="ball $img=screen_image_picker of kind $kind=spritekind || at x $x y $y"
```

Comment:
don't put the shadow in the block string. use `img.shadow="screen_image_picker"` , etc. instead. Same for the functions below this one.

Reply:
What about the kind? Is that a shadow, too?

Reply:
So, in the function below, I've got $kind=spritekind" then I've got the kind.defl=1 because I wanted it to default to a projectile, but it's not working. Should I be doing all of this differently?

Reply:
yup! anything with the `=` in the block string is a shadow. anything with parentheses after the shadow is a default value.

for example:
```
//% block="$myBall=variables_get(myBallName)"
```

should be converted to:

```
//% block="$myBall"
//% myBall.shadow=variables_get
//% myBall.defl=myBallName
```

Reply:
hmm... i'm not sure if default values work for sprite kind. try this:

```
//% kind.defl="SpriteKind.Projectile"
```

Reply:
I did…It didn't

I don't think it's a dealbreaker. Hardly anyone will probably expand.